File: Settembre/51.PhoneNumberFormatter-30Settembre.js
//todo Data una stringa di undici cifre, restituisci la stringa come numero di telefono in questo formato: "+D (DDD) DDD-DDDD".

//!

//? Tempo di esecuzione:
//* 5 Minuti e 40 secondi

//?Codice scritto senza utilizzo di ChatGpt

function formatNumber(number) {

  let prefisso = number[0]
  let numeroParentesi = number.slice(1,4)
  let preNumero = number.slice(4,7)
  let postNumero = number.slice(7)


  return `+${prefisso} (${numeroParentesi}) ${preNumero}-${postNumero}`;
}

console.log(formatNumber("05552340182"))

//?Codice scritto da ChatGpt
//* Non utilizzato

//* ✅ Pro

//* Funzione semplice e leggibile

//* Restituisce correttamente il formato desiderato

//* Usa template string in modo chiaro

//* Buona separazione delle parti del numero

//! ❌ Contro

//! Funziona solo con numeri a 11 cifre

//! Nessun controllo su input più corto o più lungo

//! Non gestisce numeri con spazi o simboli già presenti

//! Variabili poco descrittive (preNumero, postNumero)

//! Hardcoded, poco flessibile per altri formati

//? versione Chat Gpt

function formatNumber(number) {
  if (!/^\d{11}$/.test(number)) return null; // solo numeri 11 cifre //! i numeri di telefono sono sempre a 11 cifre

  const [prefisso, ...rest] = number;
  const numeroParentesi = rest.slice(0, 3).join("");
  const preNumero = rest.slice(3, 6).join("");
  const postNumero = rest.slice(6).join("");

  return `+${prefisso} (${numeroParentesi}) ${preNumero}-${postNumero}`;
}

console.log(formatNumber("05552340182"));


//todo Considerazioni: 

//! N.B.

File: Python/2.Settembre/51.Phone_number_formatter-30Settembre.py
# =========================================================
# FILE TEMPLATE - ANALISI ESERCIZI PYTHON
# Incolla direttamente questo file nel progetto VS Code
# =========================================================

# //Prompt ChatGpt: questo è il modello. Invierò consegna esercizio e codice, e vorrei feedback strutturato

# =========================================================
# //todo Consegna dell'esercizio
# =========================================================

# Data una stringa di undici cifre, restituisci la stringa
# come numero di telefono in questo formato:
#
# "+D (DDD) DDD-DDDD"

# =========================================================
# //* Codice scritto senza utilizzo di ChatGpt
# =========================================================

def format_number(number):
    prefix = number[:1]
    first_number = number[1:4]
    second_number = number[4:7]
    third_nuber = number[7:]

    return f"+{prefix} ({first_number}) {second_number}-{third_nuber}"

print(format_number("05552340182"))

# =========================================================
# //* Codice scritto da ChatGpt (versione migliorata)
# =========================================================

def format_number(number):
    return f"+{number[0]} ({number[1:4]}) {number[4:7]}-{number[7:]}"

print(format_number("05552340182"))

# =========================================================
# //* Pro e contro del mio codice
# =========================================================

# //* Pro
# - Soluzione corretta.
# - Utilizzo corretto dello slicing.
# - Codice molto leggibile.
# - Le variabili rendono chiaro il significato di ogni parte del numero.

# //! Contro
# - Sono state create variabili che vengono utilizzate una sola volta.
# - Piccolo errore di battitura nel nome "third_nuber".
# - Non viene verificato che la stringa abbia effettivamente 11 cifre.

# =========================================================
# //* Consigli
# =========================================================

# - Ottimo utilizzo dello slicing, è esattamente lo strumento giusto.
# - Fai attenzione ai nomi delle variabili per evitare typo.
# - In un contesto reale potresti aggiungere un controllo sulla lunghezza dell'input.
# - La tua soluzione è già perfettamente accettabile in un colloquio junior.

# =========================================================
# //* Valutazione
# =========================================================

# Livello: Junior
#
# Complessità temporale: O(1)
# Complessità spaziale: O(1)
#
# Passerebbe un colloquio Junior?
# Sì, senza problemi.
#
# Caso limite non gestito:
# - Stringhe con lunghezza diversa da 11.
# - Stringhe contenenti caratteri non numerici.